### ISSUE TITLE
Why does CryptTrait use the private key for encryption, rather than the public key?

### ISSUE THREAD

**Issue by alexplumb** [thephpleague/oauth2-server#684]
The CryptTrait currently gets used for encrypting/decrypting the Authorization Code and Refresh Token. Curiously, it uses the private key to encrypt and the public key to decrypt, rather than the other way around. 

This makes sense if there's an expectation for the end user to decrypt these tokens, but I can't think of any reason why I would want the sensitive information contained within these tokens to be accessible to anyone with a public key. 

If there is such an expectation, would you be open to adding a toggle in a future release so that I can choose how to encrypt this data?

*edit* Not a use case for me, but from [RFC 7636](https://tools.ietf.org/html/rfc7636#section-7.2):

> If the code challenge method is "plain" and the code challenge is to be returned inside authorization "code" to achieve a stateless server, it MUST be encrypted in such a manner that only the server can decrypt and extract it.

**Comment by alexbilbie** [comment]
I'm going to get rid of the encryption all together in the next version; it's made things overly complicated

**Comment by frederikbosch** [comment]
@alexbilbie Wrong. You should get rid of traits instead of encryption. Dependencies are not injected but with traits they are embedded, leaving the developer forced to use the CryptTrait.

You should create an interface for this. CryptInterface with two methods: encrypt and decrypt. And it should be injected in the server. Finally, your library should then provide a NoneEncryption that can be injected by default, which does no encryption at all.

This way you leave the encryption method up to the developer. Some might use OpenSSL while I rather suggest using libsodium.

Final remark: never drop security because it is overly complicated. I would urge you to leave encryption and simply come up with better code. Please, no offense, your contributions are great! But I hope you can do something with the advice.

p.s. I did a PR to make this happen a few months ago, was rejected back then because the refactoring was to big.

**Comment by alexbilbie** [comment]
I will clarify my comment as I didn't have time to yesterday; the encryption mechanism is currently used for encrypting the auth code and refresh token payloads.

In both cases I no longer feel this is necessary and representing both of these payloads as signed JWTs (as access tokens currently are represented) solves the same underlying problem and removes the home-brew encryption mechanism.

In V6 you can either use RSA signing or HMAC signing - if you're implementing a distributed system you should use RSA, if your authorization and resource server are the same codebase then use HMAC signing.

I imagine for most implementations are shared authorization and resource server codebases so I'm anticipating the implementation being *much* simpler for the vast majority of installations.

**Comment by frederikbosch** [comment]
@alexbilbie Thanks for the clarification. I hope things are implemented using interfaces rather than traits. One should only use traits as implementation of an abstraction not as the abstraction itself.

**Comment by wilsonge** [comment]
Just to add here in response to

```
This makes sense if there's an expectation for the end user to decrypt these tokens, but I can't think of any reason why I would want the sensitive information contained within these tokens to be accessible to anyone with a public key.
```

At the moment the only way for a user to check the expiry time of a refresh token out is to decrypt the token. So with the way the code base in v5 is actually I think it has to be this way around. If refresh tokens change in v6 to be full JWTs then this need goes away because the information is now easily accessible (aside: as this information is accessible it does kinda defeat the purpose you mention of this information being sensitive - this information is now totally public)

**Comment by halaei** [comment]
It sounds that encryption/decryption with public key is 50x faster than encryption/decryption with private key. So maybe that was a performance consideration: for each encryption there can be many decryption, hence the performance of decryption is more important than the performance of encryption.

Having said that, I don't know why asymmetric encryption is required.

**Comment by alexbilbie** [comment]
Asymmetric encryption allows for a better distributed architecture that doesn't require distributing the signing key material (private key) outside of the authorization server.

**Comment by alexbilbie** [comment]
This issue has been resolved as of 5.1.4 and 6.0.0
